sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/NumberUtils.scala

```diff
@@ -77,57 +101,61 @@ object NumberUtils {
       throw QueryCompilationErrors.nonFistOrLastCharInNumberFormatError(
-        s"'$letterMinusSign' or '$minusSign'", numberFormat)
-    } else if (invalidSignPosition(normalizedFormat, dollarSign)) {
+        s"'$MINUS_LETTER' or '$MINUS_SIGN'", numberFormat)
+    } else if (invalidSignPosition(normalizedFormat, DOLLAR_SIGN)) {
       throw QueryCompilationErrors.nonFistOrLastCharInNumberFormatError(
-        s"'$dollarSign'", numberFormat)
+        s"'$DOLLAR_SIGN'", numberFormat)
+    }
+  }
+
+  private def getPrecision(numberFormat: String): Int =
+    numberFormat.filterNot(isSign).length
+
+  private def getScale(numberFormat: String): Int = {
+    val formatSplits = numberFormat.split(POINT_SIGN)
+    if (formatSplits.length == 1) {
+      0
+    } else {
+      formatSplits(1).filterNot(isSign).length
     }
   }
 
   /**
    * Convert string to numeric based on the given number format.
    * The format can consist of the following characters:
-   * '9':  digit position (can be dropped if insignificant)
-   * '0':  digit position (will not be dropped, even if insignificant)
-   * '.':  decimal point (only allowed once)
-   * ',':  group (thousands) separator
-   * 'S':  sign anchored to number (uses locale)
-   * 'D':  decimal point (uses locale)
-   * 'G':  group separator (uses locale)
-   * '$':  specifies that the input value has a leading $ (Dollar) sign.
+   * '0' or '9': digit position
+   * '.' or 'D': decimal point (only allowed once)
+   * ',' or 'G': group (thousands) separator
+   * '-' or 'S': sign anchored to number
+   * '$': returns value with a leading dollar sign
    *
    * @param input the string need to converted
    * @param numberFormat the given number format
+   * @param normalizedNumberFormat normalized number format
+   * @param precision decimal precision
+   * @param scale decimal scale
    * @return decimal obtained from string parsing
    */
-  def parse(input: UTF8String, numberFormat: String): Decimal = {
-    val normalizedFormat = normalize(numberFormat)
-    check(normalizedFormat, numberFormat)
-
-    val precision = normalizedFormat.filterNot(isSign).length
-    val formatSplits = normalizedFormat.split(pointSign)
-    val scale = if (formatSplits.length == 1) {
-      0
-    } else {
-      formatSplits(1).filterNot(isSign).length
-    }
-    val transformedFormat = transform(normalizedFormat)
-    val numberFormatInstance = NumberFormat.getInstance()
-    val numberDecimalFormat = numberFormatInstance.asInstanceOf[DecimalFormat]
-    numberDecimalFormat.setParseBigDecimal(true)
-    numberDecimalFormat.applyPattern(transformedFormat)
+  private def parse(
+      input: UTF8String,
+      numberFormat: String,
+      normalizedNumberFormat: String,
+      precision: Int,
+      scale: Int): Decimal = {
     val inputStr = input.toString.trim
-    val inputSplits = inputStr.split(pointSign)
+    val inputSplits = inputStr.split(POINT_SIGN)
```

Comment:
what if the input has more than one dot?

Reply:
throws Exception.
https://github.com/apache/spark/blob/518bdad0504baa4d9af64126bd66e021959f4846/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/NumberUtils.scala#L107